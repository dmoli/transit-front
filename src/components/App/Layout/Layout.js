/* global FRONT_URL */
import Head from 'next/head';
import { defineMessages, injectIntl } from 'react-intl';
import styled from 'styled-components';

import Header from '../Header';

const messages = defineMessages({
  title: {
    id: 'page.title',
    defaultMessage: 'Transit - Smart Mobility',
  },
});

/**
 * Layout principal de la webapp
 */
export default injectIntl(({
  intl,
  title,
  children,
}) => (
  <div>
    <Head>
      <title>{ title || intl.formatMessage(messages.title) }</title>
      <meta property="og:title" content={ title || intl.formatMessage(messages.title) } />
      <meta property="og:image" content={ `${FRONT_URL}/static/logo-head.jpg` } />
      <meta property="og:description" content='Smart Mobility' />
      <meta property="og:site_name" content='Transit' />
      <meta property="og:type" content="website" />
      <meta charSet='utf-8' />
      <meta name="viewport"
        content='width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0' />
      <link href="https://fonts.googleapis.com/css?family=Poppins|Muli:300,400,700" rel="stylesheet" />
      <link rel='stylesheet' type='text/css' href='/static/css/nprogress.css' />
      <link rel='stylesheet' type='text/css' href='/static/css/style.css' />
      <link rel='apple-touch-icon' sizes='57x57' href='/apple-icon-57x57.png' />
      <link rel='apple-touch-icon' sizes='60x60' href='/apple-icon-60x60.png' />
      <link rel='apple-touch-icon' sizes='72x72' href='/apple-icon-72x72.png' />
      <link rel='apple-touch-icon' sizes='76x76' href='/apple-icon-76x76.png' />
      <link rel='apple-touch-icon' sizes='114x114' href='/apple-icon-114x114.png' />
      <link rel='apple-touch-icon' sizes='120x120' href='/apple-icon-120x120.png' />
      <link rel='apple-touch-icon' sizes='144x144' href='/apple-icon-144x144.png' />
      <link rel='apple-touch-icon' sizes='152x152' href='/apple-icon-152x152.png' />
      <link rel='apple-touch-icon' sizes='180x180' href='/apple-icon-180x180.png' />
      <link rel='icon' type='image/png' sizes='192x192' href='/android-icon-192x192.png' />
      <link rel='icon' type='image/png' sizes='32x32' href='/favicon-32x32.png' />
      <link rel='icon' type='image/png' sizes='96x96' href='/favicon-96x96.png' />
      <link rel='icon' type='image/png' sizes='16x16' href='/favicon-16x16.png' />
      <link rel='manifest' href='/manifest.json' />
      <meta name='msapplication-TileColor' content='#ffffff' />
      <meta name='msapplication-TileImage' content='/ms-icon-144x144.png' />
      <meta name='theme-color' content='#ffffff' />
    </Head>
    <Main>
      <Header />
      { children }
    </Main>
  </div>
));

const Main = styled.main`
  width: 100%;
  margin: 0 auto;
`;
