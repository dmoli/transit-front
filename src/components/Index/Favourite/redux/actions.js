import { set as setRoute } from '../../Route/redux/actions';

/**
 * Action types
 */
export const actionTypes = {
  /** Set */
  SET_FAVOURITE: 'SET_FAVOURITE',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @return action to dispatch
 */
export const set = entities => ({
  type: actionTypes.SET_FAVOURITE,
  payload: entities,
});

/**
 * Find by id
 *
 * @param {int} routeId route id
 * @param {array} routes entities
 * @return entity found
 */
const findById = (routeId, routes) => {
  const setedRoutes = JSON.parse(JSON.stringify(routes));
  let entitySeted;
  for (const entity of setedRoutes) {
    if (routeId === entity.route_id) {
      entitySeted = entity;
      break;
    }
  }
  return entitySeted;
};

/**
 * Set object fields of the array by actions
 *
 * @param {int} id id of the entity
 * @param {array} entities entities
 * @param {string} action action to do
 * @return {array} setedEntities entities seted
 */
const setFields = (id, entities, action) => {
  const setedEntities = JSON.parse(JSON.stringify(entities));
  let i = 0;
  for (const entity of setedEntities) {
    if (id === entity.route_id) {
      const entitySeted = entity;
      if (action === 'fav') entitySeted.favorited = true;
      if (action === 'unfav') entitySeted.favorited = false;
      setedEntities[i] = entitySeted;
      break;
    }
    i += 1;
  }
  return setedEntities;
};

/**
 * Add favorite - Mock only
 *
 * @param {int} routeId route id
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const add = routeId => (
  async (dispatch, getState) => {
    const state = getState();
    const route = findById(routeId, state.routes.entities);
    route.favorited = true;
    const favourites = state.favourites.entities;
    favourites.push(route);
    // add this route to state
    dispatch(set(favourites));

    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'fav');
    dispatch(setRoute(routesSeted));
  }
);

/**
 * Remove favorite - Mock only
 *
 * @param {int} routeId route id
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const remove = routeId => (
  async (dispatch, getState) => {
    const state = getState();
    let i = 0;
    const setedFavourites = JSON.parse(JSON.stringify(state.favourites.entities));
    // find route in favourite and remove it
    for (const entity of setedFavourites) {
      if (routeId === entity.route_id) {
        // delete setedFavourites[i];
        setedFavourites.splice(i, 1);
        break;
      }
      i += 1;
    }
    console.log('setedFavourites', setedFavourites);
    // refresh favourites
    dispatch(set(setedFavourites));

    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'unfav');
    dispatch(setRoute(routesSeted));
  }
);
