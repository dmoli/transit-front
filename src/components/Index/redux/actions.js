import * as api from '../../../api/routes';

/**
 * Action types
 */
export const actionTypes = {
  /** Set .entities */
  SET_ROUTES: 'SET_ROUTES',
  /** Set .entities && .page */
  SET_NEXT_PAGE_ROUTES: 'SET_NEXT_PAGE_ROUTES',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @return action to dispatch
 */
export const set = entities => ({
  type: actionTypes.SET_ROUTES,
  payload: entities,
});

/**
 * Modify entities and pagination +1
 *
 * @param {array} entities results
 * @return action to dispatch
 */
export const setNextPage = entities => ({
  type: actionTypes.SET_NEXT_PAGE_ROUTES,
  payload: entities,
});

/**
 * Get entities - API
 *
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const get = () => (
  async (dispatch) => {
    try {
      // get response - API
      const response = await api.get();
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // dispatch action
      dispatch(set(data));
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);
