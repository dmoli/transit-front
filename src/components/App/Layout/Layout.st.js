/* eslint import/no-extraneous-dependencies: [0] */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { Provider } from 'react-redux';

import Layout from './Layout';
import { initStore } from '../../../redux/store';

const store = initStore();

storiesOf('App/Layout', module)
  .addDecorator(story => <Provider store={store}>{story()}</Provider>)
  .add('Layout',
    withInfo('Layout principal de la webapp')(() =>
      <Layout title='Title' />,
    ),
  );
