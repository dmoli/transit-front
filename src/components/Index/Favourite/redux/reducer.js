import { combineReducers } from 'redux';

import initialState from '../../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.favourites.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_FAVOURITE:
      return action.payload;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
});

export default reducer;
