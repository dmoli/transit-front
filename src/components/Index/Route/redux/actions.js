/* eslint no-restricted-syntax: [0] */
import * as api from '../../../../api/routes';
import { set as setFavourite } from '../../Favourite/redux/actions';

/**
 * Action types
 */
export const actionTypes = {
  /** Set .entities */
  SET_ROUTES: 'SET_ROUTES',
  /** Set .entities && .page */
  SET_NEXT_PAGE_ROUTES: 'SET_NEXT_PAGE_ROUTES',
  /** Set .current */
  SET_CURRENT: 'SET_CURRENT',
  /** Set .shapes */
  SET_SHAPES: 'SET_SHAPES',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @param {int} page optional number page
 * @return action to dispatch
 */
export const set = (entities, page = 1) => ({
  type: actionTypes.SET_ROUTES,
  payload: {
    entities,
    page,
  },
});

/**
 * Modify entities and pagination +1
 *
 * @param {array} entities results
 * @return action to dispatch
 */
export const setNextPage = entities => ({
  type: actionTypes.SET_NEXT_PAGE_ROUTES,
  payload: entities,
});

/**
 * Modify current
 *
 * @param {string} currentName currentName
 * @return action to dispatch
 */
export const setCurrent = currentName => ({
  type: actionTypes.SET_CURRENT,
  payload: currentName,
});

/**
 * Modify shapes
 *
 * @param {array} shapes shapes
 * @return action to dispatch
 */
export const setShapes = shapes => ({
  type: actionTypes.SET_SHAPES,
  payload: shapes,
});

/**
 * Verify if a route exist in favourites
 *
 * @param {array} routes entities
 * @param {array} favourites entities
 * @return {array} entities seted
 */
const setFavoriteState = (routes, favourites) => {
  const setedRoutes = JSON.parse(JSON.stringify(routes));
  const setedFavourites = JSON.parse(JSON.stringify(favourites));
  let i = 0;
  for (const routeItem of setedRoutes) {
    for (const favouriteItem of setedFavourites) {
      if (routeItem.id === favouriteItem.id) {
        const routeSetedItem = routeItem;
        routeSetedItem.favorited = true;
        setedRoutes[i] = routeSetedItem;
      }
    }
    i += 1;
  }
  return setedRoutes;
};

/**
 * Set object fields of the array by actions
 *
 * @param {int} id id of the entity
 * @param {array} entities entities
 * @param {string} action action to do
 * @return {array} setedEntities entities seted
 */
const setFields = (id, entities, action) => {
  const setedEntities = JSON.parse(JSON.stringify(entities));
  let i = 0;
  if (action === 'current') {
    for (const entity of setedEntities) {
      const entitySeted = entity;
      if (id === entity.route_id) {
        entitySeted.current = true;
      } else {
        entitySeted.current = false;
      }
      setedEntities[i] = entitySeted;
      i += 1;
    }
  }

  i = 0;
  if (action === 'init_shape') {
    for (const entity of setedEntities) {
      const entitySeted = entity;
      entitySeted.lat = parseFloat(entitySeted.shape_pt_lat);
      entitySeted.lng = parseFloat(entitySeted.shape_pt_lon);
      setedEntities[i] = entitySeted;
      i += 1;
    }
  }
  return setedEntities;
};

/**
 * Refresh current route
 *
 * @param {int} routeId route id
 * @param {string} routeName route name
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const refreshCurrent = (routeId, routeName) => (
  async (dispatch, getState) => {
    const state = getState();
    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'current');
    dispatch(set(routesSeted, state.routes.page));

    // refresh favourites
    const favouritesSeted = setFields(routeId, state.favourites.entities, 'current');
    dispatch(setFavourite(favouritesSeted));

    // refresh current field
    dispatch(setCurrent(routeName));
  }
);

/**
 * Get entities - API
 *
 * @param {string} text text to search
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const get = (text = '', page = null) => (
  async (dispatch, getState) => {
    try {
      const state = getState();
      const pageToSearch = page === null ? state.routes.page : page;
      // get response - API
      const response = await api.get(pageToSearch, text);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // verify if a route exist in favourites
      const dataSeted = setFavoriteState(data, state.favourites.entities);
      // dispatch action
      dispatch(set(dataSeted));
      // if not exist
      if (dataSeted.length === 0) {
        throw new Error('empty');
      }

      return dataSeted;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);

/**
 * Get next page of entities - API
 *
 * @param {string} text text to search
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const getNextPage = (text = '') => (
  async (dispatch, getState) => {
    try {
      const state = getState();
      // get response - API
      const response = await api.get(state.routes.page + 1, text);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      // dispatch action
      dispatch(setNextPage(data));

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);

/**
 * Get shapes - API
 *
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const getShapes = routeId => (
  async (dispatch) => {
    try {
      // get response - API
      const response = await api.getShapes(routeId);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      // set init
      const dataSeted = setFields(null, data, 'init_shape');
      // dispatch action
      dispatch(setShapes(dataSeted));

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);
