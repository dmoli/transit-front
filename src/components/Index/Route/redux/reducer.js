import { combineReducers } from 'redux';

import initialState from '../../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.routes.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload.entities;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state.concat(action.payload);
    default:
      return state;
  }
};

/**
 * Number page reducer
 *
 * @param {array} state number page
 * @param action plus +1 in the number page
 */
const pageReducer = (state = initialState.routes.page, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload.page;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state + 1;
    default:
      return state;
  }
};

/**
 * Current name reducer
 *
 * @param {array} state current name
 * @param action change name
 */
const currentReducer = (state = initialState.routes.current, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_CURRENT:
      return action.payload;
    default:
      return state;
  }
};

/**
 * Shapes reducer
 *
 * @param {array} state list of shapes
 * @param action action to do
 */
const shapesReducer = (state = initialState.routes.shapes, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_SHAPES:
      return action.payload;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
  page: pageReducer,
  current: currentReducer,
  shapes: shapesReducer,
});

export default reducer;
