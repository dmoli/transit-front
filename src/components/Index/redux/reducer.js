import { combineReducers } from 'redux';

import initialState from '../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.routes.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state.concat(action.payload);
    default:
      return state;
  }
};

/**
 * Number page reducer
 *
 * @param {array} state number page
 * @param action plus +1 in the number page
 */
const pageReducer = (state = initialState.routes.page, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return 1;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state + 1;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
  page: pageReducer,
});

export default reducer;
