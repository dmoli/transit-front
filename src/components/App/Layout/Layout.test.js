/* eslint-env jest */
import React from 'react';
import { Provider } from 'react-redux';

import Layout from './Layout';
import { rendererWithIntl } from '../../../scripts/intl-enzyme-test-helper';
import { initStore } from '../../../redux/store';

const store = initStore();

describe('<Layout />', () => {
  it('rendered correctly', () => {
    const component = rendererWithIntl(
      <Provider store={store}>
        <Layout title='Title' />
      </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
