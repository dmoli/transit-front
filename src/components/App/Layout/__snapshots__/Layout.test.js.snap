// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> rendered correctly 1`] = `
<div>
  <main
    className="Layout__Main-yf1p0d-0 fjWrfk"
  />
</div>
`;
